forc-test/src/lib.rs

```diff
@@ -543,34 +561,66 @@ pub enum TestRunnerCount {
     Auto,
 }
 
+pub struct TestCount {
+    pub total: usize,
+    pub filtered: usize,
+}
+
+impl<'a> TestFilter<'a> {
+    fn filter(&self, fn_name: &str) -> bool {
+        if self.exact_match {
+            fn_name == self.filter_phrase
+        } else {
+            fn_name.contains(self.filter_phrase)
+        }
+    }
+}
+
 impl BuiltTests {
     /// The total number of tests.
-    pub fn test_count(&self) -> usize {
+    pub fn test_count(&self, test_filter: Option<&TestFilter>) -> TestCount {
         let pkgs: Vec<&PackageTests> = match self {
             BuiltTests::Package(pkg) => vec![pkg],
             BuiltTests::Workspace(workspace) => workspace.iter().collect(),
         };
-        pkgs.iter()
-            .map(|pkg| {
-                pkg.built_pkg_with_tests()
-                    .bytecode
-                    .entries
-                    .iter()
-                    .filter_map(|entry| entry.kind.test().map(|test| (entry, test)))
-                    .count()
-            })
-            .sum()
+        let mut num_total = 0;
+        let mut num_ignored = 0;
+        for (pkg_entry, _) in pkgs.iter().flat_map(|pkg| {
+            pkg.built_pkg_with_tests()
+                .bytecode
+                .entries
+                .iter()
+                .filter_map(|entry| entry.kind.test().map(|test| (entry, test)))
+        }) {
+            let ignored = match &test_filter {
+                Some(filter) => !filter.filter(&pkg_entry.finalized.fn_name),
+                None => false,
+            };
+            if ignored {
+                num_ignored += 1;
+            }
+            num_total += 1;
+        }
+
+        TestCount {
+            total: num_total,
+            filtered: num_ignored,
+        }
```

Comment:
nit: it'd be clearer to read if we did this whole operation with higher order functions, rather than a combination of for-loops and HOFs. This is a good use case for an accumulator.

```suggestion
        pkgs.iter()
            .flat_map(|pkg| {
                pkg.built_pkg_with_tests()
                    .bytecode
                    .entries
                    .iter()
                    .filter_map(|entry| entry.kind.test().map(|test| (entry, test)))
            })
            .fold(TestCount::default(), |acc, (pkg_entry, _)| {
                let num_ignored = match &test_filter {
                    Some(filter) => {
                        if filter.filter(&pkg_entry.finalized.fn_name) {
                            acc.filtered
                        } else {
                            acc.filtered + 1
                        }
                    }
                    None => acc.filtered,
                };
                TestCount {
                    total: acc.total + 1,
                    filtered: num_ignored,
                }
            })
```